Jbtm 3571 fix and re-enable lra quickstart

marcosgopen opened the pull request:
https://issues.redhat.com/browse/JBTM-3571

LRA quickstart updated and simplified.

marcosgopen commented:
> Pull failed: https://ci-jenkins-csb-narayana.apps.ocp-c1.prod.psi.redhat.com/job/btny-pulls-narayana-quickstart/19/

The failure is related to jakarta.ejb-api, it needs to have compile scope.

mmusgrov commented:
@marcosgopen Will you rebase this one so that we can retest it.

marcosgopen commented:
> @marcosgopen Will you rebase this one so that we can retest it.

Rebased, thanks Mike

mmusgrov commented:
The failure was caused by "wildfly-quickstart/wsat-simple does not exist" which is unrelated to this quickstart.
@marcosgopen should I go ahead and provide a review of this PR - I retested it and it looks ok now that we have fixed the LRA failure recovery of nested LRAs.

marcosgopen commented:
> The failure was caused by "wildfly-quickstart/wsat-simple does not exist" which is unrelated to this quickstart. @marcosgopen should I go ahead and provide a review of this PR - I retested it and it looks ok now that we have fixed the LRA failure recovery of nested LRAs.

Yes, please @mmusgrov . If you review it I will be able to merge this

mmusgrov commented:
I'm not seeing the second coordinator (the one in the README started with `java -Dquarkus.http.port=8081 -jar $NARAYANA_INSTALL_LOCATION/rts/lra/lra-coordinator-runner.jar`) being used. We need to check why that is before merging the PR.

marcosgopen commented:
> @mmusgrov what kind of thing would you expect to see in the log regarding seeing it being used? I can see it seems to be running on 8081 and the flight-service is seemingly configured to use it but is there something you might expect to see from the coordinator log?
> 
> I do see some ARJUNA012097 messages, are they expected? /cc @marcosgopen

Sorry Tom, I think I missed your message. Did you refer to 'WARN  [com.arj.ats.arjuna] (executor-thread-1) ARJUNA012097: Deactivate of atomic action with id 0:ffffac1c0826:8fa3:66fc333a:3 and type /StateManager/BasicAction/LongRunningAction unexpectedly failed, could not save state' ? I can see it here https://ci-jenkins-csb-narayana.apps.ocp-c1.prod.psi.redhat.com/job/btny-pulls-narayana-quickstart/382/
I'll have a look at that and let you know.

tomjenkinson commented:
Hi @marcosgopen it's a while ago but I am interested in the ARJUNA012097 messages and what you pasted looks like they have that code, thank you

mmusgrov reviewed (approved):
I can confirm that testing with what's currently in this PR is passing my crash recovery testing. I had seen the parent LRA ending up on the failed list but when I came to debug it I could no longer reproduce the failure. Neither did I see any WARNings from the coordinator. Therefore I see no reason to further delay this PR and will approve it. Marco, apologies for taking so long to retest.

For completeness, my crash recovery test was as follows:

1. Start one coordinator and the three services configured to use it.
2. Make a provisional booking but before cancelling it kill and restart the coordinator (on restart wait for it to load the LRAs from the object store).
3. Cancel the provisional booking.
4. Verify that the booking is cancelled and that the object store is empty.

mmusgrov commented:
> @mmusgrov what kind of thing would you expect to see in the log regarding seeing it being used? I can see it seems to be running on 8081 and the flight-service is seemingly configured to use it but is there something you might expect to see from the coordinator log?
> 
> I do see some ARJUNA012097 messages, are they expected? /cc @marcosgopen

ARJUNA012097 is unexpected. The message is generated when saving the LRA state [1]. BasicAction saves its part of the record [2] but it catches IOException and returns false without logging the exception. The individual participant records are saved in [3] and any IOException is logged using a WARNing. And any IOException when LongRunningAction saves its specific state is logged at WARN [4].

Therefore the warning message in the CI log is caused by ArjunaCore [2] which doesn't log the IOException. Do we need to add a new i18n log message to Core to cover this case?

[1] [https://github.com/jbosstm/narayana/blob/7.0.2.Final/rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java](https://github.com/jbosstm/narayana/blob/7.0.2.Final/rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java#L128)
[2] https://github.com/jbosstm/narayana/blob/7.0.2.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/BasicAction.java#L771
[3] https://github.com/jbosstm/narayana/blob/7.0.2.Final/rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRAParticipantRecord.java#L851
[4] https://github.com/jbosstm/narayana/blob/7.0.2.Final/rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LongRunningAction.java#L155 (edited)

marcosgopen commented:
Even if this PR is reviewed and ready I would prefer to keep it on hold an not merge it until I fixed https://issues.redhat.com/browse/JBTM-3933

tomjenkinson commented on rts/lra/README.adoc:
```diff
@@ -1,5 +1,7 @@
 ## Build the demo
 
+Make sure you export the WORKSPACE variable to the absolute path to the quickstart folder, e.g. 'export WORKSPACE=/user/myuser/quickstart/'.
```
Is it possible to get the script to fail when this is not set? Trying it here it basically seemed to hang when I did not have it set so it might be better if the script (run.sh?) checks if the variable is set and if not results in a exit -1 or something like that?

marcosgopen replied:
Thanks Tom for your suggestion, I have updated the PR.

tomjenkinson replied:
Thank you

tomjenkinson commented on rts/lra/run.sh:
```diff
@@ -59,32 +55,20 @@ function getDebugArgs {
 
-echo "Narayana installed location = $NARAYANA_INSTALL_LOCATION"
-  java ${IP_OPTS} -Dquarkus.http.port=8080 $(getDebugArgs $PORT) -jar $NARAYANA_INSTALL_LOCATION/rts/lra/lra-coordinator-runner.jar &
+  java ${IP_OPTS} -Dquarkus.http.port=8080 $(getDebugArgs $PORT) -jar $WORKSPACE/rts/lra-examples/coordinator-quarkus/target/lra-coordinator-quarkus-runner.jar &
```
This quickstart depends on the lra-examples quickstart being built first but there isn't a mention of that and it was not obvious to me. Please can I ask you to try to surface that somewhere? And also if `$WORKSPACE/rts/lra-examples/coordinator-quarkus/target/lra-coordinator-quarkus-runner.jar` does not exist please can I ask you to consider to make the script fail at that point?

marcosgopen replied:
Thanks, I have updated the script. Do you think I should also mention it in the README?

tomjenkinson replied:
Yes, I think it would be appropriate to mention something there. How about mentioning to make sure to have built $WORKSPACE/rts/lra-examples in the `Build the demo` section?

marcosgopen replied:
I put in in the 'Before you start' section, I think that is the appropriate place.

tomjenkinson replied:
Thank you!

tomjenkinson commented on rts/lra/run.sh:
```diff
@@ -1,4 +1,16 @@
+fi
+
+if [ ! -f $WORKSPACE/rts/lra-examples/coordinator-quarkus/target/lra-coordinator-quarkus-runner.jar ]; then
+    echo "Please build first the coordinator-quarkus module which is needed for this demo"
```
I think it is the `lra-coordinator-quarkus` module? It might be worth mention that this is in $WORKSPACE/rts/lra-examples

tomjenkinson commented on rts/lra/run.sh:
```diff
@@ -97,24 +93,42 @@ echo "Narayana installed location = $NARAYANA_INSTALL_LOCATION"
 
-mvn -f trip-client/pom.xml exec:java -Dexec.args="${IP_OPTS} confirm"
-mvn -f trip-client/pom.xml exec:java -Dexec.args="${IP_OPTS} cancel"
+mvn -f trip-client/pom.xml exec:java ${IP_OPTS} -Dexec.args="confirm"
```
Considering this change to IP_OPTS, I have a feeling that you need to use MAVEN_OPTS to pass in VM options? https://www.mojohaus.org/exec-maven-plugin/usage.html

marcosgopen replied:
Thanks Tom, updated: I added the line 'export MAVEN_OPTS=${IP_OPTS}' and removed the '${IP_OPTS}' from 'mvn' execution.

marcosgopen replied:
@tomjenkinson could we resolve this review?

tomjenkinson replied:
Thank you, I will resolve it

marcosgopen commented:
I verified that this quickstart works fine against lra-coordinator-quarkus 7.0.2.Final.

mmusgrov commented:
> I verified that this quickstart works fine against lra-coordinator-quarkus 7.0.2.Final.

Does this mean that we can close [JBTM-3933](https://issues.redhat.com/browse/JBTM-3933) (Nested LRA's are not cancelled when the parent cancels after coordinator restart) as not a bug?

marcosgopen commented:
> Does this mean that we can close [JBTM-3933](https://issues.redhat.com/browse/JBTM-3933) (Nested LRA's are not cancelled when the parent cancels after coordinator restart) as not a bug?

No, because it used to work with 7.0.2.Final, but not with the actual main branch. So I am looking for what was the 'breaking' change.

tomjenkinson commented:
> > Does this mean that we can close [JBTM-3933](https://issues.redhat.com/browse/JBTM-3933) (Nested LRA's are not cancelled when the parent cancels after coordinator restart) as not a bug?
> 
> No, because it used to work with 7.0.2.Final, but not with the actual main branch. So I am looking for what was the 'breaking' change.

Please can you confirm how you verified this 7.0.2.Final of the coordinator worked and please can you help me understand why the following is showing the problem anyway:

In a Narayana checkout:
```
(caution: destructve) git reset --hard 422fe15a0ef3706053403415cfa7b28f1bb58151
./build.sh -pl :lra-coordinator-war -am clean install -DskipTests
```

In a quickstart checkout at 9fa466faac7ea8c984c3a518508da87a0f31e506
```
./build.sh -pl :lra-coordinator-quarkus -am clean install -DskipTests
```
(I think this should include the LRA coordinator from above)

In a quickstart checkout at the same revision (9fa466faac7ea8c984c3a518508da87a0f31e506):
```
./build.sh -f rts/lra/lra-test/pom.xml clean install
```

Some thoughts on why this might not be doing what I think it is:
1. Possibly need to build more of Narayana than what I am above
2. Possibly need to build more of rts/lra-examples than what I am above
3. Possibly need to build more of rts/lra than what I am above with just rts/lra/lra-test

I will try full builds in the meantime...

tomjenkinson commented:
OK, so I didn't do full full builds but I think I did enough to be somewhat confident that the commit that just moves us to 7.0.3.Final-SNAPSHOT on `main` at least presents with the same problem.

I built a jbosstm/narayana working space 422fe15a0ef3706053403415cfa7b28f1bb58151 with `./build.sh clean install -DskipTests`

I build a marcosgopen/quickstart working space 9fa466faac7ea8c984c3a518508da87a0f31e506 with `/build.sh -f rts/lra-examples/pom.xml clean install` followed by `./build.sh -f rts/lra/pom.xml`

tomjenkinson commented:
I think we are in the situation where there must be a fix on `main` on the client side that is needed and a fix in `main` that is not. I can replicate the situation that Marco described with the following change:

```
diff --git a/rts/lra-examples/coordinator-quarkus/pom.xml b/rts/lra-examples/coordinator-quarkus/pom.xml
index 33bdd86aa..ac1cbbf97 100644
--- a/rts/lra-examples/coordinator-quarkus/pom.xml
+++ b/rts/lra-examples/coordinator-quarkus/pom.xml
@@ -53,6 +53,7 @@
     <dependency>
       <groupId>org.jboss.narayana.rts</groupId>
       <artifactId>lra-coordinator-jar</artifactId>
+      <version>7.0.2.Final</version>
     </dependency>
   </dependencies>
```

tomjenkinson commented:
I managed to work out basically what is going on. Without specify 7.0.2.Final we actually pull in 7.0.0.Final of the lra-coordinator-jar.

A change like this:
```
--- a/rts/lra-examples/coordinator-quarkus/pom.xml
+++ b/rts/lra-examples/coordinator-quarkus/pom.xml
@@ -53,6 +53,7 @@
     <dependency>
       <groupId>org.jboss.narayana.rts</groupId>
       <artifactId>lra-coordinator-jar</artifactId>
+      <version>7.0.3.Final-SNAPSHOT</version>
     </dependency>
   </dependencies>
```

sees the build working again. I guess 7.0.0.Final is defined in the quarkus pom in that file (https://github.com/jbosstm/quickstart/blob/main/rts/lra-examples/coordinator-quarkus/pom.xml#L31).

tomjenkinson commented:
BTW the way I eventually noticed this was using copies of a working and non-working `rts/lra-examples/coordinator-quarkus/target/lra-coordinator-quarkus-runner.jar` and `jar xf`-ing them into separate folders and `diff -r`-ing the folders and basically finding 7.0.0.Final

marcosgopen commented:
Thank you Tom very much! That explains the reason for the failure, so there is no need for https://issues.redhat.com/browse/JBTM-3933 and I am going to close it as not a bug!

tomjenkinson commented:
> Thank you Tom very much! That explains the reason for the failure, so there is no need for https://issues.redhat.com/browse/JBTM-3933 and I am going to close it as not a bug!

Thank you, Marco

marcosgopen commented:
Now I am specifying to use the project.version of all lra dependencies and there is still a problem of the recovery I am going to address tomorrow, for the moment I put the on Hold label

marcosgopen commented:
> Now I am specifying to use the project.version of all lra dependencies and there is still a problem of the recovery I am going to address tomorrow, for the moment I put the on Hold label

@tomjenkinson I addressed the recovery problem and I updated the lra quickstarts as a consequence of 'JBTM-3934 Move NarayanaLRAClient into an internal package'. Now the quickstart is ready for review.

marcosgopen commented:
I am going to create a new PR in order to update the Quarkus dependencies version.

tomjenkinson commented on rts/lra/hotel-service/pom.xml:
```diff
@@ -72,20 +71,45 @@
+    		<groupId>commons-logging</groupId>
+    		<artifactId>commons-logging</artifactId>
+    		<version>1.2</version>
+		</dependency>
```
In case you missed it, I will mention that the formatting of the dependency imports in a few places seems a bit unexpected (as in the `</dependency>` in a number of places is aligned under the previous element). Hopefully if you were to look at https://github.com/jbosstm/quickstart/pull/446/files you would see what I mean. I won't block for it but if you do decide to address it I will be happy to re-review.

marcosgopen replied:
Updated, thanks

tomjenkinson commented on rts/lra/run.sh:
```diff
@@ -59,32 +67,20 @@ function getDebugArgs {
   ID2=$!
 ((PORT++))
-  java ${IP_OPTS} -Dquarkus.http.port=8082 $(getDebugArgs $PORT) -jar hotel-service/target/quarkus-app/quarkus-run.jar &
+  java ${IP_OPTS} -Dquarkus.http.port=8082 -Dlra.http.port=8080 $(getDebugArgs $PORT) -jar hotel-service/target/quarkus-app/quarkus-run.jar &
```
@marcosgopen I don't understand why the hotel-service has a different `lra.http.port` to the flight-service but that could be my lack of familiarity with this area. I won't block on it and will assume it's intentional but mentioning this here.

marcosgopen replied:
Thanks Tom for this comment, in the demo there are two lra-coordinators and the services communicate to different lra-coordinators (hotel-service and trip-controller to the first and flight-service to the second) .  
I wanted to specify the lra.http.port for the hotel-service too for clarity. It is not important if the port is 8080 or 8081, it is good that both coordinator are in use during the demo.

tomjenkinson replied:
Thank you!

tomjenkinson commented on rts/lra/trip-controller/src/main/java/io/narayana/rts/lra/demo/tripcontroller/TripCheck.java:
```diff
@@ -37,14 +39,29 @@ static boolean validateBooking(Booking booking, boolean isConfirm, WebTarget hot
+                                    + confirmCount[0] + "\n" + booking.toJson());
+                }
+                catch (JsonProcessingException e) {
+                    // TODO Auto-generated catch block
```
I would suggest to remove the TODO here and in the next block you added but again not blocking here though

Please can I ask you to confirm that JsonProcessingException is expected to be thrown though? If it's not expected to be thrown I think you should back out that handling. I will not block on that either.

marcosgopen replied:
I'll delete the comment, that exception can be thrown by booking.toJson(), see https://github.com/jbosstm/quickstart/blob/main/rts/lra/trip-client/src/main/java/model/Booking.java#L149-L153

tomjenkinson commented on rts/lra/trip-controller/src/main/java/io/narayana/rts/lra/demo/tripcontroller/TripService.java:
```diff
@@ -37,7 +37,8 @@ public void confirmBooking(Booking booking, WebTarget hotelTarget, WebTarget fli
         lraClient.closeLRA(new URI(booking.getId()));
 
-        // Note that earlier version of the quicksart returned the new booking status in the response to the closeLRA call,
+        // Note that earlier version of the quickstart returned the new booking status
```
Personally I would remove this reference to the old way of the quickstart working and just describe how things are today but fine with me to keep it too.

marcosgopen replied:
Agreed

tomjenkinson reviewed (approved):
I added a few comments to think about but I will provide my approval as otherwise LGTM - thank you! If you decide to address any of the feedback please re-request a review from me.

marcosgopen commented on rts/lra/run.sh:
```diff
@@ -97,24 +93,43 @@ echo "Narayana installed location = $NARAYANA_INSTALL_LOCATION"
-mvn -f trip-client/pom.xml exec:java -Dexec.args="${IP_OPTS} confirm"
-mvn -f trip-client/pom.xml exec:java -Dexec.args="${IP_OPTS} cancel"
+export MAVEN_OPTS=${IP_OPTS}
+mvn -f trip-client/pom.xml exec:java -Dexec.args="confirm"
```
Updated, thanks

tomjenkinson commented on rts/lra/trip-controller/pom.xml:
```diff
@@ -89,9 +103,17 @@
+    		<groupId>commons-logging</groupId>
+    		<artifactId>commons-logging</artifactId>
+    		<version>1.2</version>
+		</dependency>
```
@marcosgopen FYI I there are a couple of misaligned </dependency> in this file but I don't think it should block it merging so I will approve the request. If you would like to fix it, please do re-request a review but otherwise I am good with it being merged. Thank you!

tomjenkinson reviewed (approved):
There is a little bit of misaligned XML I pointed out but I would not block merging for those so I am going to approve - thank you!

tomjenkinson reviewed (approved):
LGTM - thank you!
